sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Load test for {@link ParDo} operation.
+ *
+ * <p>The purpose of this test is to measure {@link Combine}'s behaviour in stressful conditions. It
+ * uses {@link SyntheticBoundedIO} and {@link SyntheticStep} which both can be parametrized to
+ * generate keys and values of various size, impose delay (sleep or cpu burnout) in various moments
+ * during the pipeline execution and provide some other performance challenges.
+ *
+ * @see SyntheticStep
+ * @see SyntheticBoundedIO
+ *     <p>You can choose between multiple combine modes to test per key combine operations ({@link
+ *     CombinerType}).
+ *     <p>To run it manually, use the following command:
+ *     <pre>
+ *    ./gradlew :beam-sdks-java-load-tests:run -PloadTest.args='
+ *      --fanout=1
+ *      --perKeyCombinerType=TOP_LARGEST
+ *      --topCount=10
+ *      --sourceOptions={"numRecords":1000,...}
+ *      --stepOptions={"outputRecordsPerInputRecord":2...}'
+ *      -PloadTest.mainClass="org.apache.beam.sdk.loadtests.CombineLoadTest"
+ * </pre>
+ */
+public class CombineLoadTest extends LoadTest<CombineLoadTest.Options> {
+
+  private static final String METRICS_NAMESPACE = "combine";
+
+  private enum CombinerType {
+    TOP_LARGEST,
+    MEAN
+  }
+
+  /** Pipeline options specific for this test. */
+  interface Options extends LoadTestOptions {
+
+    @Description("Number consequent of ParDo operations (SyntheticSteps) to be performed.")
+    @Default.Integer(1)
+    Integer getNumberOfCounterOperations();
+
+    void setNumberOfCounterOperations(Integer count);
+
+    @Description("The number of Combine operations to perform in parallel.")
+    @Default.Integer(1)
+    Integer getFanout();
+
+    void setFanout(Integer fanout);
+
+    @Description("Per key combiner type.")
+    @Default.Enum("MEAN")
+    CombinerType getPerKeyCombinerType();
+
+    void setPerKeyCombinerType(CombinerType combinerType);
+
+    @Description("Number of top results to combine (if applicable).")
+    Integer getTopCount();
+
+    void setTopCount(Integer topCount);
+  }
+
+  private CombineLoadTest(String[] args) throws IOException {
+    super(args, Options.class, METRICS_NAMESPACE);
+  }
+
+  @Override
+  protected void loadTest() throws IOException {
+    Combine.PerKey combiner = createPerKeyCombiner(options.getPerKeyCombinerType());
+
+    Optional<SyntheticStep> syntheticStep = createStep(options.getStepOptions());
+
+    PCollection<KV<byte[], byte[]>> input =
+        pipeline
+            .apply("Read input", SyntheticBoundedIO.readFrom(sourceOptions))
+            .apply("Collect metrics", ParDo.of(new MetricsMonitor(METRICS_NAMESPACE)));
+
+    for (int i = 0; i < options.getFanout(); i++) {
+      applyStepIfPresent(input, format("Step: %d", i), syntheticStep)
+          .apply(
+              format("Convert to BigInteger: %d", i), MapElements.via(new ByteValueToBigInteger()))
+          .apply(format("Combine: %d", i), combiner);
```

Comment:
I need some feedback about what Combine modes should we test. Are Top.perKeyLargest and Mean.perKey combine operations enough?

Reply:
Let's add sum and count.

Reply:
ok